docs/source/datasets.rst

```diff
@@ -3,6 +3,51 @@ torchtext.datasets
 
 .. currentmodule:: torchtext.datasets
 
+
+.. _datapipes_warnings:
+
+.. warning::
+
+    The datasets supported by torchtext are datapipes from the `torchdata
+    project <https://pytorch.org/data/beta/index.html>`_, which is still in Beta
+    status. This means that the API is subject to change without deprecation
+    cycles. In particular, we expect a lot of the current idioms to change with
+    the eventual release of ``DataLoaderV2`` from ``torchdata``.
+
+    Here are a few recommendations regarding the use of datapipes:
+
+    - For shuffling the datapipe, do that in the DataLoader: ``DataLoader(dp, shuffle=True)``.
+      You do not need to call ``dp.shuffle()``, because ``torchtext`` has
+      already done that for you. Note however that the datapipe won't be
+      shuffled unless you explicitly pass ``shuffle=True`` to the DataLoader.
+
+    - When using multi-processing (``num_workers=N``), use the builtin ``worker_init_fn``::
+
+            from torch.utils.data.backward_compatibility import worker_init_fn
+            DataLoader(dp, num_workers=4, worker_init_fn=worker_init_fn, drop_last=True)
+
+      This will ensure that data isn't duplicated across workers.
+
+    - We also recommend using ``drop_last=True``. Without this, the batch sizes
+      at the end of an epoch may be very small in some cases (smaller than with
+      other map-style datasets). This might affect accuracy greatly especially
+      when batch-norm is used. ``drop_last=True`` ensures that all batch sizes
+      are equal.
+
+    - Distributed training with ``DistributedDataParallel`` is not yet entirely
+      stable / supported, and we don't recommend it at this point. It will be
+      better supported in DataLoaderV2. If you still wish to use DDP, make sure
+      that:
+
+      - All workers (DDP workers *and* DataLoader workers) see a different part
+        of the data. You might need to call ``dp = dp.apply_sharding(world_size,
+        rank)``.
+      - The shuffling seed is the same across all workers. You might need to
+        call ``torch.utils.data.graph_settings.apply_shuffle_seed(dp, rng)``
+      - The suffling seed is different across epochs.
+      - The rest of the RNG (typically used for transformations) is
+        **different** across workers, for maximal entropy and optimal accuracy.
+
```

Comment:
It's not unusual to have different number of samples in each rank, for example, if the datapipe expects the same number of tokens/characters for each batch, each sentence may have different number of tokens/characters, then the batch size will be different. As long as the loss value is normalized in a correct way, the training should be fine.

Reply:
@NicolasHug Oh you are referring the number of batches in each rank. That's correct, DDP will hang forever if the number of batches are not even in all ranks.  I was talking about the batch_size and it can be different :)

Reply:
Thanks all, I added a section about all this just above

>  - All DDP workers work on the same number of batches. One way to do this is to by limit the size of the datapipe within each worker to ``len(datapipe) // num_ddp_workers``, but this might not suit all use-cases.

Reply:
yupp, SGTM!